enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/protocol/handshake/HandshakeClient.java

```diff
@@ -19,42 +19,68 @@
+import org.neo4j.stream.Streams;
 
 import static org.neo4j.causalclustering.protocol.handshake.StatusCode.SUCCESS;
 
 // TODO: modifier protocols
 public class HandshakeClient implements ClientMessageHandler
 {
     private Channel channel;
-    private ProtocolRepository protocolRepository;
-    private ProtocolSelection knownProtocolVersions;
+    private ProtocolRepository<Protocol.ApplicationProtocol> applicationProtocolRepository;
+    private ProtocolRepository<Protocol.ModifierProtocol> modifierProtocolRepository;
+    private ProtocolSelection<Protocol.ApplicationProtocol> knownApplicationProtocolVersions;
+    private Set<ProtocolSelection<Protocol.ModifierProtocol>> knownModifierProtocolVersions;
+    private Protocol.ApplicationProtocol applicationProtocol;
+    private Map<String,Optional<Protocol.ModifierProtocol>> negotiatedModifierProtocols = new HashMap<>();
     private ProtocolStack protocolStack;
     private CompletableFuture<ProtocolStack> future = new CompletableFuture<>();
     private boolean magicReceived;
 
-    public CompletableFuture<ProtocolStack> initiate( Channel channel, ProtocolRepository protocolRepository, Protocol.Identifier protocol )
+    public CompletableFuture<ProtocolStack> initiate( Channel channel, ProtocolRepository<Protocol.ApplicationProtocol> applicationProtocolRepository,
+            Protocol.ApplicationProtocolIdentifier applicationProtocolIdentifier, ProtocolRepository<Protocol.ModifierProtocol> modifierProtocolRepository,
+            Set<Protocol.ModifierProtocolIdentifier> modifierProtocolIdentifiers )
     {
         this.channel = channel;
-        this.protocolRepository = protocolRepository;
-        this.knownProtocolVersions = protocolRepository.getAll( protocol );
+        this.applicationProtocolRepository = applicationProtocolRepository;
+        this.knownApplicationProtocolVersions = applicationProtocolRepository.getAll( applicationProtocolIdentifier );
 
-        channel.write( new InitialMagicMessage() );
-        channel.writeAndFlush( new ApplicationProtocolRequest( knownProtocolVersions.identifier(), knownProtocolVersions.versions() ) );
+        this.modifierProtocolRepository = modifierProtocolRepository;
+        this.knownModifierProtocolVersions = modifierProtocolIdentifiers
+                .stream()
+                .map( modifierProtocolRepository::getAll )
+                .collect( Collectors.toSet() );
+
+        channel.write( InitialMagicMessage.instance() );
+
+        knownModifierProtocolVersions.forEach( modifierProtocolSelection ->
+                channel.write( new ModifierProtocolRequest( modifierProtocolSelection.identifier(), modifierProtocolSelection .versions() ) ) );
+
+        channel.writeAndFlush( new ApplicationProtocolRequest( knownApplicationProtocolVersions.identifier(), knownApplicationProtocolVersions.versions() ) );
 
         return future;
     }
 
-    private void ensureMagic()
+    private void ensureMagic( Runnable block )
```

Comment:
This seems more obfuscated compared to the previous version where the pre-conditions are checked and the method aborts (throws) early when they aren't fulfilled. It creates more entangling between two pieces of functionality that could be quite separate concerns.

Reply:
The HandshakeServer threw on a missing magic but HandshakeClient did not - it exceptionally completed its future and then happily proceeded to send messages to the server. Maybe throwing would be better here. I don't think coupling is an issue though - ensureMagic just knows it should do what it's told if the magic value has been received.